### ISSUE TITLE
Overwrite a nested class in derived class 

### ISSUE THREAD

**Issue by enhuiz** [microsoft/pyright#2627]
**Describe the bug**
When a nested class is overwritten in the derived class,  pyright seems not aware of this change and treats the nested class as the original one.

**To Reproduce**
```python
class A:
    class Data:
        x = 1

    def __init__(self):
        self.data = self.Data()


class B(A):
    class Data:
        y = 2


print(A().data.x)
print(type(A().data))

print(B().data.x)
print(B().data.y)
print(type(B().data))

```

**Expected behavior**
Expect x is unknown and y is known.

**Screenshots or Code**
![image](https://user-images.githubusercontent.com/12977366/143732280-61329469-aeac-4555-a80d-133ba2676595.png)

**VS Code extension or command-line**
nvim with coc-pyright.

**Comment by erictraut** [comment]
I think pyright's behavior is correct in this case, so I don't consider this a bug. The `data` instance variable has no type annotation, so its type is inferred to be of type `A.Data` based on the assignment.

This is a very unusual pattern, not one that I've seen before. It is legal Python code, but it's not very conducive to static type checking. If your intent is for `data` to refer to a different class for each subclass of `A`, you'll need to use generics or protocols to type it correctly. I can't tell enough from your sample code how you plan to use the `Data` class. If you would like to provide additional details, I'm happy to suggest ways that you could annotate it appropriately.

**Comment by erictraut** [comment]
The most straightforward annotation suggestion I can think of is something like the following:

```python
class A:
    data: "A.Data"

    class Data:
        x = 1

    def __init__(self):
        self.data = self.Data()

class B(A):
    data: "B.Data"

    class Data:
        y = 2
```

**Comment by enhuiz** [comment]
Thanks, @erictraut. your suggested way seems good for me.

I was trying to do something like this:

```python
class AData:
    x = 1


class A:
    def __init__(self):
        self.data = self.Data()

    @property
    def Data(self):
        return AData


class BData(AData):
    y = 2


class B(A):
    @property
    def Data(self):
        return BData
```

where the main class and its data class are separately inherited so that the data can be managed properly without messing up with the member variables from the main class. I nested the data class into the main class in the above example. I just noticed that it may not be good practice because I changed the type of `Data` in the derived class.
